### DISCUSSION TITLE
Can zserio use both big and small endian byte order?

### DISCUSSION THREAD

**Discussion by kancve** [ndsev/zserio#590] [Ideas]
Recently, I want to use zserio to parse [WKB](https://www.osgeo.cn/doc_ogcstd/ogc_standard/ch02_chapter1/chapter.html#wkb) format streams, and I found that WKB streams can specify either of two byte orders, but zserio only using big endian byte order. What modifications do zserio need to make to support it?

Attached is the content I defined:

![wkb](https://github.com/ndsev/zserio/assets/29945891/35a31f15-9e80-48b7-b8b8-2c3c052829e4)

Best wishes!

**Comment by mikir** [comment]
Hello @kancve!

Thanks that you are still using zserio! Unfortunatelly, you were hit by one Zserio drawback. Because there is no header in Zserio serialized binary data, there is no possibility where to store byte order of upcoming data.

However, you can probably do it by yourself using some tricks in schema. Example:

```
struct Data
{
    bool isBigIndian;
    ValueU16(isBigIndian) value;
};

struct ValueU16(bool isBigIndian)
{
    uint16 origValue;
    
    function uint16 convertToHost()
    {
        return (isBigIndian) ? origValue : ((origValue & 0xFF) << 8) | ((origValue >> 8) & 0xFF);
    }
};
```

Then, application must access the field `value` using `convertToHost()` function only. Probably, other or better solution could be found. This is just the first idea that came to my mind.

**Comment by mikir** [comment]
Regarding possible solution how to support big/little endian directly in Zserio, please see #124.

**Comment by kancve** [comment]
Hello @mikir!
Thank you for your prompt attention. This trick in schema seems ineffective for float types as it cannot directly perform displacement operations.

I am also seeking a solution and researching how [GDAL](https://gdal.org/drivers/vector/gpkg.html) handles it. I hope to bring inspiration for the improvement of zserio. 
Thanks again.

**Comment by mikir** [comment]
Thanks a lot for your effort. Any inspiration will be pretty welcome!

PS.: I will convert this issue to the discussion.
